OpenRA.Mods.Common/Traits/BotModules/UnitBuilderBotModule.cs

```diff
@@ -0,0 +1,203 @@
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Controls AI unit production.")]
+	public class UnitBuilderBotModuleInfo : ConditionalTraitInfo
+	{
+		public readonly HashSet<string> McvTypes = new HashSet<string>();
+		public readonly HashSet<string> ConstructionYardTypes = new HashSet<string>();
+		public readonly HashSet<string> McvFactoryTypes = new HashSet<string>();
+		public readonly HashSet<string> RefineryTypes = new HashSet<string>();
+		public readonly HashSet<string> PowerTypes = new HashSet<string>();
+
+		// TODO: Investigate whether this might the (or at least one) reason why bots occasionally get into a state of doing nothing.
+		// Reason: If this is less than SquadSize, the bot might get stuck between not producing more units due to this,
+		// but also not creating squads since there aren't enough idle units.
+		[Desc("Only produce units as long as there are less than this amount of units idling inside the base.")]
+		public readonly int IdleBaseUnitsMaximum = 12;
+
+		[Desc("Production queues AI uses for producing units.")]
+		public readonly HashSet<string> UnitQueues = new HashSet<string> { "Vehicle", "Infantry", "Plane", "Ship", "Aircraft" };
+
+		[Desc("What units to the AI should build.", "What relative share of the total army must be this type of unit.")]
+		public readonly Dictionary<string, int> UnitsToBuild = null;
+
+		[Desc("What units should the AI have a maximum limit to train.")]
+		public readonly Dictionary<string, int> UnitLimits = null;
+
+		public override object Create(ActorInitializer init) { return new UnitBuilderBotModule(init.Self, this); }
+	}
+
+	public class UnitBuilderBotModule : ConditionalTrait<UnitBuilderBotModuleInfo>, IBotTick, IBotIdleBaseUnits
+	{
+		public const int FeedbackTime = 30; // ticks; = a bit over 1s. must be >= netlag.
+
+		readonly World world;
+		readonly Player player;
+
+		int ticks;
+		int unitsHangingAroundBaseCount;
+
+		public UnitBuilderBotModule(Actor self, UnitBuilderBotModuleInfo info)
+			: base(info)
+		{
+			world = self.World;
+			player = self.Owner;
+		}
+
+		void IBotIdleBaseUnits.UpdatedUnitsHangingAroundBaseCount(int newCount)
+		{
+			unitsHangingAroundBaseCount = newCount;
+		}
+
+		void IBotTick.BotTick(IBot bot)
+		{
+			if (IsTraitDisabled)
+				return;
+
+			ticks++;
+
+			if (ticks % FeedbackTime == 0)
+				ProductionUnits(bot);
+		}
+
+		void ProductionUnits(IBot bot)
+		{
+			// Stop building until economy is restored
+			if (!HasAdequateRefineryCount)
+				return;
+
+			// No construction yards - Build a new MCV
+			if (Info.McvTypes.Any() && HasAdequateConstructionYardCount &&
```

Comment:
Can we please generalize this by defining an `IBotRequestUnitProduction` interface implemented by this module which adds the unit type to a queue, and then builds those here instead of hardcoding MCV?

We can then split out a separate `McvManagerBotModule` to manage MCV production and deployment logic.